cmd/object-api-utils.go

```diff
@@ -842,16 +844,21 @@ func newS2CompressReader(r io.Reader) io.ReadCloser {
 	return pr
 }
 
-// Returns error if the cancelCh has been closed (indicating that S3 client has disconnected)
-type detectDisconnect struct {
+// Returns error if the context is canceled, indicating
+// either client has disconnected
+type contextReader struct {
 	io.ReadCloser
-	cancelCh <-chan struct{}
+	ctx context.Context
 }
 
-func (d *detectDisconnect) Read(p []byte) (int, error) {
+func (d *contextReader) Read(p []byte) (int, error) {
 	select {
-	case <-d.cancelCh:
-		return 0, io.ErrUnexpectedEOF
+	case <-d.ctx.Done():
+		err := d.ctx.Err()
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			return 0, io.ErrUnexpectedEOF
```

Comment:
I generally don't like rewriting errors. This makes it impossible to detect if it was *actually* an `io.ErrUnexpectedEOF` or the context was cancelled or exceed the timeout.

If this is done to correct some error elsewhere I would rather have that be fixed.

Reply:
It isn't just that returning this error sort of makes sense for the usage of this reader, so that the caller fails appropriately, let me check if we handle context errors as.

Reply:
I changed the error to be handled by the caller @klauspost